fix: multiple changes in the Jupyter notebooks

gnikit opened the pull request:
## Changes

- Replaces references to `code` to `src`
- Adds language attributes to fenced code blocks
- Updates output of `setup.cfg` in notebook 2.
- Fixes spelling and grammar errors
- Standardised the README using the [template](https://raw.githubusercontent.com/ImperialCollegeLondon/ReCoDE-README-template/main/README-template.md?token=GHSAT0AAAAAABPRNXTZCOF7UC74ZT5T5P5YYWWUL3A)
- Edits the Introduction to include instructions on how to clone the repo
- Adds Learning Outcomes
- Renames the conda venv from `recode` to `mcmc` to avoid conflicts with other (future and present) ReCoDE exemplars

gnikit committed:
fix: name conflict with other exemplars

The venv cannot be named recode, since a student might be working
through multiple exemplars which would lead to name conflicts
for the venvs.

I named the conda venv mcmc for Monte Carlo Markov Chain.

gnikit committed:
fix: multiple changes in the Jupyter notebooks

- Replaces references to `code` to `src`
- Adds language attributes to fenced code blocks
- Updates output of `setup.cfg` in notebook 2.
- Fixes spelling and grammar errors

gnikit committed:
docs: fix math rendering in sphinx website

TomHodson commented:
Looks good! Can you add a little section mentioning how to enable Mathjax to the documentation chapter?

gnikit commented:
> Looks good! Can you add a little section mentioning how to enable Mathjax to the documentation chapter?

Sure, for .rst files or for Jupyter notebooks, parsed with sphinx? For some reason the latter requires you to dig through the settings of `myst-nb` instead of working out of the box. In contrast, [nbsphinx](https://nbsphinx.readthedocs.io/en/0.8.9/) just worked by I thought I would stick with the Jupyter parser you chose.

gnikit commented:
@TomHodson I think we should move the API docs section at the end of the Table of Contents. As it currently stands it interferes with the flow of the exemplar. Quickstart --> Intro --> ... seems much smoother than Quickstart --> API Docs --> Intro -->...

gnikit committed:
docs: add descriptions to sphinx exts + mathajax

gnikit committed:
chore: fix string formatting

gnikit committed:
docs: move API docs at the end of the exemplar

gnikit commented:
@TomHodson Is there somewhere a list of the Learning Outcomes of the exemplar?

gnikit committed:
docs: Add GitHub clone instructions at Intro

gnikit committed:
chore: updated link to ImperialCollegeLondon

Also removed duplicate link to Jupyter notebooks

gnikit committed:
docs:  edited README to adhere to template

Template can be found in:
https://raw.githubusercontent.com/ImperialCollegeLondon/ReCoDE-README-template/main/README-template.md?token=GHSAT0AAAAAABPRNXTZCOF7UC74ZT5T5P5YYWWUL3A

gnikit commented:
This is now done, please have a look at the updated PR description.

TomHodson commented:
> @TomHodson I think we should move the API docs section at the end of the Table of Contents. As it currently stands it interferes with the flow of the exemplar. Quickstart --> Intro --> ... seems much smoother than Quickstart --> API Docs --> Intro -->...

Yeah I agree

TomHodson commented:
> @TomHodson Is there somewhere a list of the Learning Outcomes of the exemplar?

No but Jeremy and I can come up with some tomorrow

TomHodson commented:
All looks good!

jcohen02 commented on README.md:
```diff
@@ -16,35 +16,64 @@
+When you're ready to dive in you have 4 options:
 
-### 1. Launch them in Binder (easiest but a bit slow)
+### 1. Launch them in Binder
```
Suggesting a slight modification here for clarity:
```suggestion
### 1. Launch the notebooks in Binder
```

jcohen02 commented on README.md:
````diff
@@ -53,9 +82,18 @@ pip install .[dev]
 ```
 
-### 3. View them non-interactively in GitHub via the links in the table of contents
+_NOTE: Better performance but requires you have python and Jupyter installed_.
````
Very minor point but trying to ensure consistency in terms of whether Python is capitalised:
```suggestion
_NOTE: Better performance but requires you have Python and Jupyter installed_.
```

jcohen02 commented on docs/learning/01 Introduction.ipynb:
```diff
@@ -12,20 +12,27 @@
+    "\n",
+    "or if for some reason you can't do that (because you are on a phone or tablet for instance) you can instead open this notebook in [binder](https://mybinder.org/v2/gh/TomHodson/ReCoDE_MCMCFF/HEAD)\n",
+    "\n",
+    "I would also suggest you set up a Python environment just for this. You can use your preferred method to do this, but I will recommend `Anaconda` because it's both what I currently use and what is recommended by Imperial.\n",
```
Add "project":
```suggestion
    "I would also suggest you set up a Python environment just for this project. You can use your preferred method to do this, but I will recommend `Anaconda` because it's both what I currently use and what is recommended by Imperial.\n",
```

jcohen02 commented on docs/learning/01 Introduction.ipynb:
```diff
@@ -73,15 +80,15 @@
    "metadata": {},
    "source": [
-    "We're going to be working with arrays of numbers so it will make sense to work with `Numpy` and we'll also want to plot things, the standard choice for this is `matplotlib`, though there are other options, `pandas` and `plotly` being notable ones.\n",
+    "We're going to be working with arrays of numbers, so it will make sense to work with `Numpy`, and we'll also want to plot things, the standard choice for this is `matplotlib`, though there are other options, `pandas` and `plotly` being notable ones.\n",
```
NumPy instead of Numpy? (see https://numpy.org/) 🙂  (plus a couple of other suggested changes while I'm editing this...)
```suggestion
    "We're going to be working with arrays of numbers, so it will make sense to work with `NumPy`, and we'll also want to plot things, the standard choice for this is `Matplotlib`, though there are other options, `pandas` and `Plotly` being notable ones.\n",
```

jcohen02 commented on docs/learning/01 Introduction.ipynb:
```diff
@@ -389,25 +396,25 @@
     "\n",
     "#### Using Arrays\n",
-    "One way is we work with arrays of numbers and operations defined over those arrays such as `sum`, `product` etc. `Numpy` is the canonical example of this in Python but many machine learning libraries are essentually doing a similar thing. We rely on the library to implement the operations efficiently and try to chain those operations together to achieve what we want. This imposes some limita ...
+    "One way is we work with arrays of numbers and operations defined over those arrays such as `sum`, `product` etc. `Numpy` is the canonical example of this in Python, but many machine learning libraries are essentially doing a similar thing. We rely on the library to implement the operations efficiently and try to chain those operations together to achieve what we want. This imposes some limit ...
```
NumPy:
```suggestion
    "One way is we work with arrays of numbers and operations defined over those arrays such as `sum`, `product` etc. `NumPy` is the canonical example of this in Python, but many machine learning libraries are essentially doing a similar thing. We rely on the library to implement the operations efficiently and try to chain those operations together to achieve what we want. This imposes some limitations on the way we can write our code.\n",
```

jcohen02 commented on docs/learning/01 Introduction.ipynb:
```diff
@@ -613,7 +620,7 @@
     "So far we've discussed the problem we want to solve, written a little code, tested it a bit and made some speed improvements.\n",
     "\n",
-    "In the next notebook we will package the code up into a little python package, this is has two big benefits to use: \n",
+    "In the next notebook we will package the code up into a little python package, this has two big benefits to use: \n",
```
Suggesting a slight rewording:
```suggestion
    "In the next notebook we will package the code up into a little python package, this has two big benefits when using the code: \n",
```

jcohen02 commented on docs/learning/02 Packaging It Up.ipynb:
```diff
@@ -30,9 +30,9 @@
     "\n",
     "\n",
-    "Before we can do any testing, it is best practice to structure and then package your code up as a python project up. You don't have to do it like this but but it carrys with it the benefit that many only tutorial _expect_ you to do it like this and generally you want to reduce friction for yourself later. \n",
+    "Before we can do any testing, it is best practice to structure and then package your code up as a python project up. You don't have to do it like this, but it carries with it the benefit that many other tutorials _expect_ you to do it like this, and generally you want to reduce friction for yourself later. \n",
```
typo fixed:
```suggestion
    "Before we can do any testing, it is best practice to structure and then package your code up as a python project. You don't have to do it like this, but it carries with it the benefit that many other tutorials _expect_ you to do it like this, and generally you want to reduce friction for yourself later. \n",
```

jcohen02 commented on docs/learning/02 Packaging It Up.ipynb:
````diff
@@ -53,15 +53,15 @@
     "```\n",
     "\n",
-    "It's looks pretty intimidating! But let's quickly go through it, at the top level of most projects you'll find on Github and elsewhere you'll find files to do with the project as a whole:\n",
+    "It's looks pretty intimidating! But let's quickly go through it, at the top level of most projects you'll find on GitHub, and elsewhere you'll find files to do with the project as a whole:\n",
````
I may be misreading this but it doesn't seem to read correctly - I'm suggesting a rewording but feel free to ignore if I've misunderstood! 🙂 
```suggestion
    "It looks pretty intimidating! But let's quickly go through it: at the top level of most projects you'll find on GitHub (and elsewhere) there are a group of files that describe the project as a whole or provide key project information - not all projects will have all of these files and, indeed, there a variety of other files that you may also see so this is an example of some of the more important files:\n",
```

jcohen02 commented on docs/learning/02 Packaging It Up.ipynb:
```diff
@@ -53,15 +53,15 @@
     "- `README.md` - An intro to the project\n",
     "- `LICENSE` - The software license that governs this project, there are a few standard ones people use.\n",
-    "- `environment.yaml` (or both) this list what python packages the project needs in a standard format\n",
+    "- `environment.yml` (or both) this list what python packages the project needs in a standard format\n",
```
```suggestion
    "- `environment.yml` (or alternatives) - this lists what Python packages the project needs in a standard format (other languages have equivalents).\n",
```

jcohen02 commented on docs/learning/02 Packaging It Up.ipynb:
```diff
@@ -53,15 +53,15 @@
+    "Then below that you will usually have directories breaking the project up into main categories, here I have `src/` and `docs/learning/`.\n",
     "\n",
-    "Inside `code/` we have a standard python package directory structure.\n",
+    "Inside `src/` we have a standard python package directory structure.\n",
```
```suggestion
    "Inside `src/` we have a standard Python package directory structure.\n",
```

jcohen02 commented on docs/learning/02 Packaging It Up.ipynb:
```diff
@@ -195,9 +190,9 @@
  "metadata": {
   "kernelspec": {
-   "display_name": "Python [conda env:recode]",
+   "display_name": "Python 3.8.10 ('venv': venv)",
```
I guess this got changed because you ran it locally? I had the same thing with my edits - not a big issue but I wonder if this should be reverted to Tom's original so it ties in with his use of a conda environment?
```suggestion
   "display_name": "Python [conda env:recode]",
```

jcohen02 commented on docs/learning/02 Packaging It Up.ipynb:
```diff
@@ -195,9 +190,9 @@
+   "display_name": "Python 3.8.10 ('venv': venv)",
    "language": "python",
-   "name": "conda-env-recode-py"
+   "name": "python3"
```
See previous comment...
```suggestion
   "name": "conda-env-recode-py"
```

jcohen02 commented on docs/learning/02 Packaging It Up.ipynb:
```diff
@@ -209,7 +204,12 @@
+  "vscode": {
+   "interpreter": {
+    "hash": "f5403acae4671aac0ae5a29dd5903d33d0105a9e9d4148f755d3321f5023d387"
+   }
```
As before, see previous comment...
```suggestion
   "version": "3.9.12"
```

jcohen02 commented on docs/learning/03 Writing a Markov Chain Monte Carlo Sampler.ipynb:
```diff
@@ -48,14 +48,14 @@
     "\n",
     "\n",
-    "For our Ising model the equivalent question would be what's the average color of this system? At high temperatures we expect the pixels to be random and average out ot grey, while at low temperatures they will all be either black or while.\n",
+    "For our Ising model the equivalent question would be what's the average color of this system? At high temperatures we expect the pixels to be random and average out out grey, while at low temperatures they will all be either black or while.\n",
```
typos:
```suggestion
    "For our Ising model the equivalent question would be what's the average color of this system? At high temperatures we expect the pixels to be random and average out grey, while at low temperatures they will all be either black or white.\n",
```

jcohen02 commented on docs/learning/03 Writing a Markov Chain Monte Carlo Sampler.ipynb:
```diff
@@ -139,9 +139,9 @@
    "metadata": {},
    "source": [
-    "These images give a flavour of why physicists find this model useful, it gives window into how thermal noise and spontaneous order interact. At low temperatures the energy cost of being different from your neighbours is the most important thing, while at high temperatures, it doesn't matter and you really just do your own thing.\n",
+    "These images give a flavour of why physicists find this model useful, it gives window into how thermal noise and spontaneous order interact. At low temperatures the energy cost of being different from your neighbours is the most important thing, while at high temperatures, it doesn't matter, and you really just do your own thing.\n",
```
typo:
```suggestion
    "These images give a flavour of why physicists find this model useful, it gives a window into how thermal noise and spontaneous order interact. At low temperatures the energy cost of being different from your neighbours is the most important thing, while at high temperatures, it doesn't matter, and you really just do your own thing.\n",
```

jcohen02 commented on docs/learning/03 Writing a Markov Chain Monte Carlo Sampler.ipynb:
```diff
@@ -139,9 +139,9 @@
+    "These images give a flavour of why physicists find this model useful, it gives window into how thermal noise and spontaneous order interact. At low temperatures the energy cost of being different from your neighbours is the most important thing, while at high temperatures, it doesn't matter, and you really just do your own thing.\n",
     "\n",
-    "There's a special point somewhere in the middle called the critical point $T_c$ where all sorts of cool things happen, but my favourite is that for large system sizes you get a kind of fractal behaviour which I will demonstrate more once we've sped this code up and can simulate larger systems in a reasonable time. You can kinda see it for 50x50 systesm at T = 5 but not really clearly."
+    "There's a special point somewhere in the middle called the critical point $T_c$ where all sorts of cool things happen, but my favourite is that for large system sizes you get a kind of fractal behaviour which I will demonstrate more once we've sped this code up and can simulate larger systems in a reasonable time. You can kinda see it for 50x50 system at T = 5 but not really clearly."
```
typo:
```suggestion
    "There's a special point somewhere in the middle called the critical point $T_c$ where all sorts of cool things happen, but my favourite is that for large system sizes you get a kind of fractal behaviour which I will demonstrate more once we've sped this code up and can simulate larger systems in a reasonable time. You can kinda see it for a 50x50 system at T = 5 but not really clearly."
```

jcohen02 commented on docs/learning/03 Writing a Markov Chain Monte Carlo Sampler.ipynb:
```diff
@@ -206,9 +206,9 @@
  "metadata": {
   "kernelspec": {
-   "display_name": "Python [conda env:recode]",
+   "display_name": "Python 3.8.10 ('venv': venv)",
```
```suggestion
   "display_name": "Python [conda env:recode]",
```

jcohen02 commented on docs/learning/03 Writing a Markov Chain Monte Carlo Sampler.ipynb:
```diff
@@ -206,9 +206,9 @@
+   "display_name": "Python 3.8.10 ('venv': venv)",
    "language": "python",
-   "name": "conda-env-recode-py"
+   "name": "python3"
```
```suggestion
   "name": "conda-env-recode-py"
```

jcohen02 commented on docs/learning/03 Writing a Markov Chain Monte Carlo Sampler.ipynb:
```diff
@@ -220,7 +220,12 @@
+  "vscode": {
+   "interpreter": {
+    "hash": "f5403acae4671aac0ae5a29dd5903d33d0105a9e9d4148f755d3321f5023d387"
+   }
```
```suggestion
   "version": "3.9.12"
```

jcohen02 commented on docs/learning/05 Adding Functionality.ipynb:
```diff
@@ -40,7 +40,7 @@
     "# Adding Functionality\n",
     "\n",
-    "The main thing we want to be able to do is to take measurements, the code as I have writting it doesn't really allow that because it only returns the final state in the chain. Let's say we have a measurement called `average_color(state)` that we want to average over the whole chain. We could just stick that inside our definition of `mcmc` but we know that we will likely make other measuremen ...
+    "The main thing we want to be able to do is to take measurements, the code as I have writing it doesn't really allow that because it only returns the final state in the chain. Let's say we have a measurement called `average_color(state)` that we want to average over the whole chain. We could just stick that inside our definition of `mcmc`, but we know that we will likely make other measuremen ...
```
typo fixed plus minor suggested edits:
```suggestion
    "The main thing we want to be able to do is to take measurements. The code, as I have written it, doesn't really allow that because it only returns the final state in the chain. Let's say we have a measurement called `average_color(state)` that we want to average over the whole chain. We could just stick that inside our definition of `mcmc`, but we know that we will likely make other measurements too, and we don't want to keep writing new versions of our core functionality!\n",
```

jcohen02 commented on docs/learning/05 Adding Functionality.ipynb:
```diff
@@ -52,11 +52,11 @@
     "### Option 1: Just save all the states and return them\n",
     "\n",
-    "The problem with this is the states are very big and we don't want to waste all that memory. For an NxN state that uses 8 bit integers (the smallest we can use in numpy) 1000 samples would already use 2.5Gb of memory! We will see later that we'd really like to be able to go a bit bigger than 50x50 and 1000 samples!\n",
+    "The problem with this is the states are very big, and we don't want to waste all that memory. For an `NxN` state that uses 8-bit integers (the smallest we can use in numpy) `1000` samples would already use `2.5Gb` of memory! We will see later that we'd really like to be able to go a bit bigger than `50x50` and `1000` samples!\n",
```
Replace Gb (gigabits) with GB (gigabytes) since I assume the latter is what's meant. 
```suggestion
    "The problem with this is the states are very big, and we don't want to waste all that memory. For an `NxN` state that uses 8-bit integers (the smallest we can use in NumPy) `1000` samples would already use `2.5GB` (2.5 gigabytes) of memory! We will see later that we'd really like to be able to go a bit bigger than `50x50` and `1000` samples!\n",
```

jcohen02 commented on docs/learning/05 Adding Functionality.ipynb:
````diff
@@ -73,7 +73,7 @@
     "```\n",
     "\n",
-    "This could work but it limits how we can store measurements and what shape and type they can be. What if we want to store our measurements in a numpy array? Or what if your measurement itself is a vector or and object that can't easily be stored in a numpy array? We would have to think carefully about what functionality we want."
+    "This could work, but it limits how we can store measurements and what shape and type they can be. What if we want to store our measurements in a numpy array? Or what if your measurement itself is a vector or and object that can't easily be stored in a numpy array? We would have to think carefully about what functionality we want."
````
Typo (plus changing numpy -> NumPy while I'm suggesting a fix):
```suggestion
    "This could work, but it limits how we can store measurements and what shape and type they can be. What if we want to store our measurements in a NumPy array? Or what if your measurement itself is a vector or an object that can't easily be stored in a NumPy array? We would have to think carefully about what functionality we want."
```

jcohen02 commented on docs/learning/05 Adding Functionality.ipynb:
```diff
@@ -153,7 +153,7 @@
    "metadata": {},
    "source": [
-    "This requires only a very small change to our mcmc function and suddenly we can do whatever we like with the states! While we're at it I'm going to add an aditional argument `stepsize` that allows us to only sample the state every `stepsize` MCMC steps. You'll see why we would want to set this to value greater than 1 in a moment."
+    "This requires only a very small change to our `mcmc` function, and suddenly we can do whatever we like with the states! While we're at it, I'm going to add an argument `stepsize` that allows us to only sample the state every `stepsize` MCMC steps. You'll see why we would want to set this to value greater than 1 in a moment."
```
```suggestion
    "This requires only a very small change to our `mcmc` function, and suddenly we can do whatever we like with the states! While we're at it, I'm going to add an argument `stepsize` that allows us to only sample the state every `stepsize` MCMC steps. You'll see why we would want to set this to a value greater than 1 in a moment."
```

jcohen02 commented on docs/learning/06 Speeding It Up.ipynb:
```diff
@@ -40,7 +40,7 @@
     "# Speeding It Up\n",
     "\n",
-    "In order to show you a really big system will still need to make the code a bit faster. Right now we calculate the energy of each state, flip a pixel and then calculate the energy again. It turns out that you can actually directly calculate the energy change instead of doing this subtraction. Let's do this is a sort of test driven decelopment fashion: we want to write a function that when gi ...
+    "In order to show you a really big system will still need to make the code a bit faster. Right now we calculate the energy of each state, flip a pixel and then calculate the energy again. It turns out that you can actually directly calculate the energy change instead of doing this subtraction. Let's do this is a sort of test driven development fashion: we want to write a function that when gi ...
```
typos:
```suggestion
    "In order to show you a really big system, we will still need to make the code a bit faster. Right now we calculate the energy of each state, flip a pixel, and then calculate the energy again. It turns out that you can actually directly calculate the energy change instead of doing this subtraction. Let's do this in a sort of test-driven development fashion: we want to write a function that when given a state and a pixel to flip, returns how much the energy goes up by (negative if down) upon performing the flip.\n",
```

I also note that this mentions test-driven development - we can't assume readers will know what this means - I think anything that people aren't familiar with risks making them confused -- if TDD will already have been mentioned at this point, that's fine, otherwise we could maybe add a note here (even if it just says "don't worry if you don't know what test-driven development is. If you're not familiar with it and would like to find out more, take a look [here]" or something to that effect?

jcohen02 commented on docs/learning/07 Producing Research Outputs.ipynb:
```diff
@@ -40,7 +40,7 @@
     "# Producing Research Outputs\n",
     "\n",
-    "So now that we have the ability to simulate our system lets do a little exploration. First let's take three temperatures at each we'll do 10 runs and see how the systems evolve. I'll also tack on a little histogram at the right hand side of where the systens spent their time."
+    "So now that we have the ability to simulate our system let's do a little exploration. First let's take three temperatures at each we'll do `10` runs and see how the systems evolve. I'll also tack on a little histogram at the right-hand side of where the systems spent their time."
```
I found this a little confusing to read - suggesting minor edits to aid readability:
```suggestion
    "So now that we have the ability to simulate our system let's do a little exploration. First let's take three temperatures. For each we'll do `10` runs and see how the systems evolve. I'll also tack on a little histogram at the right-hand side showing where the systems spent their time."
```

jcohen02 commented on docs/learning/07 Producing Research Outputs.ipynb:
```diff
@@ -138,9 +138,9 @@
-    "- At intermediate temperature the fluctuations occur on much longer time scales! Because the systems can only move a little bit each timestep, it means that the measurements we are making are *correlated* with themselves at previous times. The result of this is that if we use MCMC to draw N samples, we don't get as much information as if we had drawn samples from an uncorrelated variable (li ...
+    "- It takes a while for MCMC to 'settle in', you can see that for T = 10 the natural state is somewhere around c = 0, which takes about 2000 steps to reach from the initial state with c = 1. In general when doing MCMC we want to throw away some values at the beginning because they're too affected by the initial state.\n",
+    "- At High and Low temperatures we basically just get small fluctuations about an average value\n",
+    "- At intermediate temperature the fluctuations occur on much longer time scales! Because the systems can only move a little each timestep, it means that the measurements we are making are *correlated* with themselves at previous times. The result of this is that if we use MCMC to draw N samples, we don't get as much information as if we had drawn samples from an uncorrelated variable (like a ...
```
Suggesting some minor updates:
(also, maybe say "influenced" rather than "affected" below?)
```suggestion
    "- It takes a while for MCMC to 'settle in', you can see that for T = 10 the natural state is somewhere around c = 0, which takes about 2000 steps to reach from the initial state with c = 1. In general when doing MCMC we want to throw away some values at the beginning because they're affected too much by the initial state.\n",
    "- At High and Low temperatures we basically just get small fluctuations around an average value\n",
    "- At intermediate temperatures the fluctuations occur on much longer time scales! Because the systems can only move a little each timestep, it means that the measurements we are making are *correlated* with themselves at previous times. The result of this is that if we use MCMC to draw N samples, we don't get as much information as if we had drawn samples from an uncorrelated variable (like a die roll for instance)."
```

jcohen02 commented on docs/learning/08 Doing Reproducible Science.ipynb:
```diff
@@ -15,15 +15,15 @@
    "source": [
     "# Doing Reproducible Science\n",
-    "Further Reading on this software reproducability: [The Turing Way: Guide to producing reproducable research](https://the-turing-way.netlify.app/reproducible-research/reproducible-research.html)\n",
+    "Further Reading on this software reproducibility: [The Turing Way: Guide to producing reproducible research](https://the-turing-way.netlify.app/reproducible-research/reproducible-research.html)\n",
```
```suggestion
    "Further reading on the reproducibility of software outputs: [The Turing Way: Guide to producing reproducible research](https://the-turing-way.netlify.app/reproducible-research/reproducible-research.html)\n",
```

jcohen02 commented on docs/learning/08 Doing Reproducible Science.ipynb:
```diff
@@ -15,15 +15,15 @@
+    "There are many aspects to this, but I'll list what I think is relevant here:\n",
     "1. We have some code that generates the data and some code that uses it to plot the output, let's split that into two python files.\n",
-    "2. Our code has external dependencies on numpy and matplotlib, in the future those packages could change their behaviour in a way that breaks the code, so lets record what version our code is compatible with.\n",
+    "2. Our code has external dependencies on numpy and matplotlib, in the future those packages could change their behaviour in a way that breaks the code, so let's record what version our code is compatible with.\n",
```
```suggestion
    "2. Our code has external dependencies on `numpy` and `matplotlib`. In the future those packages could change their behaviour in a way that breaks (our changes the output of) our code, so let's record what version our code is compatible with.\n",
```

jcohen02 commented on docs/learning/08 Doing Reproducible Science.ipynb:
```diff
@@ -15,15 +15,15 @@
+    "2. Our code has external dependencies on numpy and matplotlib, in the future those packages could change their behaviour in a way that breaks the code, so let's record what version our code is compatible with.\n",
     "3. We also have an internal dependency on other code in this MCFF repository, that could also change so let's record the git hash of the commit where the code works for posterity.\n",
-    "4. The data generating process is random so we'll fix the seed as discussed in the testing section to make it reproducable.\n",
+    "4. The data generating process is random, so we'll fix the seed as discussed in the testing section to make it reproducible.\n",
```
```suggestion
    "4. The data generation process is random, so we'll fix the seed as discussed in the testing section to make it reproducible.\n",
```

jcohen02 commented on docs/learning/08 Doing Reproducible Science.ipynb:
````diff
@@ -112,8 +126,9 @@
    "source": [
-    "So I also output `conda env export`, this is annoying because it also gives you dependencies. While the versions of dependencies coudl potentially be important we usually draw the line at just listing the version of directly required packages. So what I usually do is to take the above output and then use the the output of `conda env export` to set the version numbers, leaving out the number  ...
-    "```\n",
+    "So I also output `conda env export`, this is annoying because it also gives you dependencies. While the versions of dependencies could potentially be important we usually draw the line at just listing the version of directly required packages. So what I usually do is to take the above output and then use the output of `conda env export` to set the version numbers, leaving out the number beca ...
````
I found this a little confusing - I presume by "directly required packages" it's referring to Python packages used directly by the MCFF codebase? I'd class these as dependencies. When it then talks about dependencies, I think it's actually referring to "dependencies of dependencies", i.e. the full dependency tree? I wonder if this could be made a little bit clearer. Apologies if I've misunderstood.

gnikit replied:
That's what I understood too. @TomHodson is it okay if this gets slightly reworded?

jcohen02 commented on docs/learning/08 Doing Reproducible Science.ipynb:
```diff
@@ -137,13 +152,13 @@
     "To avoid you having to go away and find the files, I'll just put them here. Let's start with the file that generates the data. I'll give it what I hope is an informative name and a shebang so that we can run it with `./generate_montecarlo_walkers.py` (after doing `chmod +x generate_montecarlo_walkers.py` just once).\n",
     "\n",
-    "I'll set the seed using a large pregenerated seed, you've likely seen me use 42 in some places but that's not really best practive because it might not be entropy to reliably seed the generator.\n",
+    "I'll set the seed using a large pregenerated seed, you've likely seen me use `42` in some places, but that's not really best practice because it might not be entropy to reliably seed the generator.\n",
```
```suggestion
    "I'll set the seed using a large pregenerated seed, you've likely seen me use `42` in some places, but that's not really best practice because it might not provide good enough entropy to reliably seed the generator.\n",
```

jcohen02 commented on docs/learning/08 Doing Reproducible Science.ipynb:
```diff
@@ -372,11 +387,11 @@
+    "Now that we have a nicely reproducible plot, let's share it with the world. The easiest way is probably to put your code in a hosted git repository like GitHub or GitLab. \n",
     "\n",
-    "Next, let's mint a shiny Digital Object Identifier (DOI) for the repository, using something like [Zenodo](https://zenodo.org/). These services archive a snapshot of the repository and assign a DOI to that snapshot, this is realy useful for citing a particular version of the software. \n",
+    "Next, let's mint a shiny Digital Object Identifier (DOI) for the repository, using something like [Zenodo](https://zenodo.org/). These services archive a snapshot of the repository and assign a DOI to that snapshot, this is really useful for citing a particular version of the software. \n",
```
```suggestion
    "Next, let's mint a shiny Digital Object Identifier (DOI) for the repository, using something like [Zenodo](https://zenodo.org/). These services archive a snapshot of the repository and assign a DOI to that snapshot, this is really useful for citing a particular version of the software, e.g. in a publication (and helping to ensure that published results are reproducible by others). \n",
```

jcohen02 commented on docs/learning/08 Doing Reproducible Science.ipynb:
```diff
@@ -372,11 +387,11 @@
+    "Next, let's mint a shiny Digital Object Identifier (DOI) for the repository, using something like [Zenodo](https://zenodo.org/). These services archive a snapshot of the repository and assign a DOI to that snapshot, this is really useful for citing a particular version of the software. \n",
     "\n",
-    "Finally let's add a `citation.cff` file to the root of the repository, this makes it easier for people who might cite this software to generate a good citation for it. We can add the zenodo DOI to it too. You can read more about `citation.cff` files [here](https://citation-file-format.github.io/) and there is a convenient generator tool [here](https://citation-file-format.github.io/cff-initi ...
+    "Finally, let's add a `citation.cff` file to the root of the repository, this makes it easier for people who might cite this software to generate a good citation for it. We can add the Zenodo DOI to it too. You can read more about `citation.cff` files [here](https://citation-file-format.github.io/) and there is a convenient generator tool [here](https://citation-file-format.github.io/cff-init ...
```
```suggestion
    "Finally, let's add a `CITATION.cff` file to the root of the repository, this makes it easier for people who want to cite this software to generate a good citation for it. We can add the Zenodo DOI to it too. You can read more about `CITATION.cff` files [here](https://citation-file-format.github.io/) and there is a convenient generator tool [here](https://citation-file-format.github.io/cff-initializer-javascript/)."
```

jcohen02 commented on docs/learning/09 Adding Documentation.ipynb:
```diff
@@ -30,7 +30,7 @@
     "We'll use sphinx along with a couple plugins: [autodoc][autodoc] allows us to generate documentation automatically from the docstrings in our source code, while [napoleon][napoleon] allows us to use [NUMPYDOC][numpydoc] and Google formats for the docstrings in addition to [reStructuredText][rst]\n",
     "\n",
-    "What this means is that we'll be able to write documentation directly into the source code and it will get rendered into a nice website. This helps keep the documentation up to date beause it's right there next to the code!\n",
+    "What this means is that we'll be able to write documentation directly into the source code, and it will get rendered into a nice website. This helps keep the documentation up to date because it's right there next to the code!\n",
```
```suggestion
    "What this means is that we'll be able to write documentation directly into the source code and it will get rendered into a nice website. This helps keep the documentation up to date because it's right there next to the code and the web-based documentation will get automatically re-generated every time the documentation files are updated!\n",
```

jcohen02 commented on docs/learning/09 Adding Documentation.ipynb:
```diff
@@ -173,7 +175,7 @@
     "### Documentation Ideas\n",
     "\n",
-    "Readthedocs can be a bit tricky to setup, it is also possible to use Github pages to acomplish something similar. Another idea is to include some simple copyable code snippets in a quickstart guide. This lets people get up and running your code more quickly than is they need to read the API docs to understand how to interact with your module."
+    "Readthedocs can be a bit tricky to set up, it is also possible to use GitHub pages to accomplish something similar. Another idea is to include some simple copyable code snippets in a quickstart guide. This lets people get up and running your code more quickly than is they need to read the API docs to understand how to interact with your module."
```
Suggesting slight rewording/typo fixed:
```suggestion
    "Readthedocs can be a bit tricky to set up, it is also possible to use [GitHub Pages](https://pages.github.com/) to accomplish something similar. Another idea is to include some simple copyable code snippets in a quickstart guide. This lets people get up and running with your code more quickly than if they need to read the API documentation to understand how to interact with your module."
```

jcohen02 commented on docs/learning/04 Testing.ipynb:
```diff
@@ -25,7 +25,7 @@
     "Ok we can finally start writing and running some tests!\n",
     "\n",
-    "I copied some of the initial tests that we did in chapter 1 into `test_energy.py` installed pytest into my development environment with `pip install pytest`. If you're using conda you need to use `conda install pytest` and now I can run the `pytest` command in the ReCoDE_MCFF directory. Pytest will automatically discover our tests and run them, to do this it relies on their being python file ...
+    "I copied some of the initial tests that we did in chapter 1 into `test_energy.py` installed pytest into my development environment with `pip install pytest`. If you're using conda you need to use `conda install pytest`, and now I can run the `pytest` command in the `mcmc` directory. Pytest will automatically discover our tests and run them, to do this it relies on their being python files wi ...
```
typo:
```suggestion
    "I copied some of the initial tests that we did in chapter 1 into `test_energy.py` and installed pytest into my development environment with `pip install pytest`. If you're using conda you need to use `conda install pytest`. I can now run the `pytest` command in the `mcmc` directory. Pytest will automatically discover our tests and run them. To do this it relies on there being Python files with functions named `test_\\*` which it will run. It's also a widely used convention to begin the name of Python files containing tests with `test_`\n",
```

jcohen02 commented on docs/learning/04 Testing.ipynb:
````diff
@@ -92,19 +92,19 @@
     "```\n",
     "\n",
-    "I will defer to external resources for a full discussion of the philosphy of testing but I generally think of tests as an aid to my future debugging. If I make a change that breaks something then I want my tests to catch that and to make it clear what has broken. As such I generally put tests that check very basic properties of my code early on in the file and then follow them with tests tha ...
+    "I will defer to external resources for a full discussion of the philosophy of testing, but I generally think of tests as an aid to my future debugging. If I make a change that breaks something then I want my tests to catch that and to make it clear what has broken. As such I generally put tests that check very basic properties of my code early on in the file and then follow them with tests t ...
````
typo:
```suggestion
    "I will defer to external resources for a full discussion of the philosophy of testing, but I generally think of tests as an aid to my future debugging. If I make a change that breaks something then I want my tests to catch that and to make it clear what has broken. As such I generally put tests that check very basic properties of my code early on in the file and then follow them with tests that probe more subtle things or more obscure edge cases.\n",
```

jcohen02 commented on docs/learning/04 Testing.ipynb:
```diff
@@ -116,7 +116,7 @@
     "## Coverage Testing\n",
     "\n",
-    "A useful little trick for testing, are tools like pytest-cov that can measure *coverage*, that is, the amount of your code base that is activate by your tests. Unfortunatley Numba does not play super well with pytest-cov so we have to turn off numba to generate the test report using an environment variable.\n",
+    "A useful little trick for testing, are tools like pytest-cov that can measure *coverage*, that is, the amount of your code base that is activated by your tests. Unfortunately Numba does not play super well with pytest-cov, so we have to turn off numba to generate the test report using an environment variable.\n",
```
I don't think I'd call coverage a "trick", it's a pretty fundamental element of testing - maybe we can slightly edit this?! 🙂 
```suggestion
    "A useful aspect of testing is *coverage*. This involves looking at how much of your code is actually "covered" by the tests you've written. That is, which individual lines of your code are actually being run by your tests. Tools like `pytest-cov` can measure _coverage_. Unfortunately Numba does not play super well with `pytest-cov`, so we have to turn off Numba using an environment variable so that we can run `pytest-cov` and generate the "test report".\n",
```

jcohen02 commented on docs/learning/04 Testing.ipynb:
````diff
@@ -144,7 +144,7 @@
     "```\n",
     "\n",
-    "Ok so this is telling us that we currently test 86% of the lines in ising_model.py. We can also change `--cov-report=html` to get a really nice html output which shows which parts of your code aren't being run.\n",
+    "Ok so this is telling us that we currently test 86% of the lines in ising_model.py. We can also change `--cov-report=html` to get a really nice `html` output which shows which parts of your code aren't being run.\n",
````
```suggestion
    "Ok so this is telling us that we currently test 86% of the lines in ising_model.py. We can also change `--cov-report=html` to get a really nice `html` output which shows which parts of our code aren't being run.\n",
```

jcohen02 commented on docs/learning/04 Testing.ipynb:
```diff
@@ -156,7 +156,7 @@
     "## Advanced Testing Methods: Property Based Testing\n",
     "\n",
-    "I won't do into huge detail here but I thought it would be nice to make you aware of a nice library called `Hypothesis` that helps with this problem of finding edge cases. `Hypothesis` gives you tools to generate randomised inputs to functions, so as long as you can come up with some way to verify the output is correct or has the correct _properties_ (or just that the code doens't throw and  ...
+    "I won't do into huge detail here, but I thought it would be nice to make you aware of a nice library called `Hypothesis` that helps with this problem of finding edge cases. `Hypothesis` gives you tools to generate randomised inputs to functions, so as long as you can come up with some way to verify the output is correct or has the correct _properties_ (or just that the code doesn't throw and ...
```
typo + link added:
```suggestion
    "I won't go into huge detail here, but I thought it would be nice to make you aware of a nice library called [`Hypothesis`](https://hypothesis.readthedocs.io) that helps with this problem of finding edge cases. `Hypothesis` gives you tools to generate randomised inputs to functions, so as long as you can come up with some way to verify the output is correct or has the correct _properties_ (or just that the code doesn't throw and error!) then this can be a powerful method of testing. \n",
```

jcohen02 commented on docs/learning/04 Testing.ipynb:
````diff
@@ -169,7 +169,7 @@
     "    assert np.allclose(energy(state), energy_numpy(state))\n",
     "```\n",
-    "You tell Hypothesis how to generate the test data, in this case we use some numpy specifc code to generate 2 dimensional arrays with `dtype = int` and entries randomly sampled from `[1, -1]`. We use the same trick as before of checking two implementations against one another."
+    "You tell Hypothesis how to generate the test data, in this case we use some numpy specific code to generate 2 dimensional arrays with `dtype = int` and entries randomly sampled from `[1, -1]`. We use the same trick as before of checking two implementations against one another."
````
```suggestion
    "You tell Hypothesis how to generate the test data. In this case we use some NumPy specific code to generate 2 dimensional arrays with `dtype = int` and entries randomly sampled from `[1, -1]`. We use the same trick as before of checking two implementations against one another."
```

jcohen02 commented on docs/learning/04 Testing.ipynb:
```diff
@@ -181,14 +181,14 @@
     "## Testing Stochastic Code\n",
     "\n",
-    "We have a interesting problem here, most testing assumes that for the same inputs we will always get the same outputs but our MCMC sampler is a stochastic algorithm. So how can we test it? I can see three mains routes we can take:\n",
+    "We have an interesting problem here, most testing assumes that for the same inputs we will always get the same outputs but our MCMC sampler is a stochastic algorithm. So how can we test it? I can see three mains routes we can take:\n",
```
typo:
```suggestion
    "We have an interesting problem here, most testing assumes that for the same inputs we will always get the same outputs but our MCMC sampler is a stochastic algorithm. So how can we test it? I can see three main routes we can take:\n",
```

jcohen02 commented on docs/learning/04 Testing.ipynb:
```diff
@@ -357,11 +357,11 @@
     "1. We are taking 2000 samples of a random variable X, those samples has some mean $m$ and standard deviation $\\sigma_X$, the mean is the center of mass of the above histogram and the standard deviation is a measure of how wide it is.\n",
     "\n",
-    "2. However what we actually want to do is ask \"How close is the mean to 0?\", to answer that we need to know how much we expect the mean to vary by when we rerun the calculation. Turns the mean of N samples of a variable X then the mean varies by \n",
+    "2. However, what we actually want to do is ask \"How close is the mean to 0?\", to answer that we need to know how much we expect the mean to vary by when we rerun the calculation. Turns the mean of N samples of a variable X then the mean varies by \n",
```
What is meant in the second sentence here? "Turns"? I think this is a typo or there's a word missing?!

gnikit replied:
My guess would be that it's meant to be "Turns out"?

gnikit replied:
@TomHodson can you have a look?

jcohen02 replied:
@TomHodson feel free to suggest an alternative, however, looking again, I'm assuming that this is intended to say:
```suggestion
    "2. However, what we actually want to do is ask \"How close is the mean to 0?\", to answer that we need to know how much we expect the mean to vary by when we rerun the calculation. It turns out that given N samples of a variable X, the mean varies by \n",
```

jcohen02 commented on docs/learning/04 Testing.ipynb:
```diff
@@ -383,9 +383,9 @@
     "## Test Driven Development\n",
     "\n",
-    "I won't talk about TDD much here but it's likely a term you will hear at some point. It essentially referes to the practice of writing tests as part of your process of writing code. Rather than writing all your code and then writing tests for them. You could instead write some or all of your tests upfront and then write code that passes them. \n",
+    "I won't talk about TDD much here, but it's likely a term you will hear at some point. It essentially refers to the practice of writing tests as part of your process of writing code. Rather than writing all your code and then writing tests for them. You could instead write some or all of your tests upfront and then write code that passes them. \n",
```
```suggestion
    "I won't talk much about Test Driven Development, or TDD, here, but it's likely a term you will hear at some point. It essentially refers to the practice of writing tests as part of your process of writing code. Rather than writing all your code and then writing tests for them. You could instead write some or all of your tests upfront, describing the expected bahviour of code that doesn't yet exist, and then write the necessary code so that your tests pass. \n",
```

jcohen02 commented on docs/learning/04 Testing.ipynb:
````diff
@@ -403,21 +403,21 @@
     "    -   id: black-jupyter\n",
     "```\n",
-    "And finally `pre-commit install` will make this run every time you commit to git. It's worth running it manually once the first time to check it works: `pre-commit run --all-files`. Running this I immediatly got a cryptic error that, on googling, turned out to be that something broke in version 21.12b0 of `21.12b0`. Running `precommit autoupdate` fixed this for me by updated `black` to a lat ...
+    "And finally `pre-commit install` will make this run every time you commit to git. It's worth running it manually once the first time to check it works: `pre-commit run --all-files`. Running this I immediately got a cryptic error that, on googling, turned out to be that something broke in version 21.12b0 of `21.12b0`. Running `precommit autoupdate` fixed this for me by updated `black` to a la ...
````
typo:
```suggestion
    "And finally `pre-commit install` will make this run every time you commit to git. It's worth running it manually once the first time to check it works: `pre-commit run --all-files`. Running this I immediately got a cryptic error that, on googling, turned out to be that something broke in version 21.12b0 of `black`. Running `precommit autoupdate` fixed this for me by updating `black` to a later version. Running `pre-commit run --all-files` a second time now gives me:\n",
```

jcohen02 reviewed (changes requested):
Great work @gnikit (and, of course, fantastic job on preparing this resource @TomHodson), many thanks for going through this in such detail and making so many comments/updates.

I've gone through your comments and made a number of suggestions along the way. Some are where I've spotted additional typos or minor grammatical issues, others are where I felt something in the original text wasn't completely clear or was difficult to read and I've suggested some alternative wording. Feel free to accept or reject my suggestions as you see best.

I'll mark this as "Request changes" for now and can take a quick final re-check once you've had a chance to go through my comments if that would be useful. Otherwise feel free to go ahead and merge when you're ready (and let me know if you need me to give a final approval on the PR before you're able to merge).

gnikit committed:
Apply suggestions from code review

Co-authored-by: Jeremy Cohen <jcohen02@users.noreply.github.com>

gnikit commented:
Thanks @jcohen02 for the review! I have gone ahead and committed all the requested changes. It would appear that my "smart" machine learning spell checker is not that smart after all.

Two things that we might want to address before we merge are:

- [ ] collapsable sections in Jupyter notebooks for the solutions, haven't found any great solutions to this yet. Jupyter book, has an implementation but that would require swapping the Sphinx docs, with Jupyter book which I would not recommend.
- [x] Making sure that the Binder link works. I have been trouble getting the environment to initialise. It could be just me, but it would be good if we could verify.

jcohen02 reviewed (approved):
Thanks for running through the comments and suggestions @gnikit. I've suggested alternative wording for one of the queries where we're awaiting input from @TomHodson but I'll leave that with Tom to check.

Re your two questions at the end:
 - Binder: the "launch binder" badge in the README works fine for me. I click it, I see the starting up page and it takes a couple of minutes to spin up a server and then it drops me straight into a jupyter instance in the browser showing the notebooks.
 - Collapsible sections: I don't know the answer to this either - I guess you've probably looked through all the possible options but I can take a quick look anyway.

I'd be happy for this to be merged and look at collapsible sections as a separate item (if there isn't already an issue for them, we should open one). Equally, if you prefer to hold off on merging and try to resolve that issue within this PR, also fine from my side.

Either way, I'm happy to approve this now.

TomHodson commented:
I wouldn’t have any problem with you swapping nbsphinx in. the fact that
myst doesn’t support Mathjax by default is pretty weird.

On Tue, 19 Jul 2022 at 09:27, gnikit ***@***.***> wrote:

> Looks good! Can you add a little section mentioning how to enable Mathjax
> to the documentation chapter?
>
> Sure, for .rst files or for Jupyter notebooks, parsed with sphinx? For
> some reason the latter requires you to dig through the settings of myst-nb
> instead of working out of the box. In contrast, nbsphinx
> <https://nbsphinx.readthedocs.io/en/0.8.9/> just worked by I thought I
> would stick with the Jupyter parser you chose.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/ImperialCollegeLondon/ReCoDE_MCMCFF/pull/5#issuecomment-1188755190>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAPX4SCABEYSU4M7662ZDW3VUZRH5ANCNFSM53572MRA>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

TomHodson commented:
Happy for you to merge this when ready.

TomHodson commented:
Ah I see there are some comments waiting my input, I'll take a look when I'm on my desktop.

I do find binder a bit hit and miss, they cache the environments but delete them after a period of time so if you spin one up when the environment has aged out of the cache it takes absolutely ages to initialise.

gnikit commented:
> I wouldn’t have any problem with you swapping nbsphinx in. the fact that myst doesn’t support Mathjax by default is pretty weird.

It's working now so I wouldn't bother. I think it's more like, myst-nb does not support `$$` by default. My guess is that they are doing so to avoid triggering the Mathjax parser erroneously.

I can commit the suggested changes for you if you want.

gnikit committed:
chore: renamed table header

gnikit commented:
@TomHodson good to merge this one?

TomHodson commented:
Yeah!

Did that typo that Jeremy pointed out make its way into this? Or should I click the "commit suggestion" button? 

For the collapsible answers, Jing's suggestion to use HTML [summary tags](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/summary) could work but it means the use has to copy past the code to run it, so the notebooks wouldn't actually run to completion without manual intervention.

jcohen02 commented:
@TomHodson, re collapsible cells with answers, does this help by any chance? [https://github.com/aquirdTurtle/Collapsible_Headings](https://github.com/aquirdTurtle/Collapsible_Headings). Assuming there's a way to tell some content to be collapsed by default, I wonder if this might give you what you want?

gnikit commented:
> Yeah!
> 
> Did that typo that Jeremy pointed out make its way into this? Or should I click the "commit suggestion" button?

I think you need to commit the changes for those.

> 
> For the collapsible answers, Jing's suggestion to use HTML [summary tags](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/summary) could work but it means the use has to copy past the code to run it, so the notebooks wouldn't actually run to completion without manual intervention.

I would leave the collapsible sections for the another PR. I think we've added too many changes on this one. I am of the opinion that the Jupyter notebooks should be able to run, so the HTML solution is not idea.

TomHodson committed:
Apply suggestions from code review

Co-authored-by: Jeremy Cohen <jcohen02@users.noreply.github.com>

gnikit commented:
> @TomHodson, re collapsible cells with answers, does this help by any chance? https://github.com/aquirdTurtle/Collapsible_Headings. Assuming there's a way to tell some content to be collapsed by default, I wonder if this might give you what you want?

There are some other extensions out there that might be easier to use:

- https://jupyter-contrib-nbextensions.readthedocs.io/en/latest/nbextensions/collapsible_headings/readme.html
- https://stackoverflow.com/questions/33159518/collapse-cell-in-jupyter-notebook
